### PR TITLE
fcmp++: fix path_for_proof when a non-root layer has single elem

### PR DESCRIPTION
**Pull request by j-berman**
This edge case was triggered by the tree being in a particular state (check the test case for a deeper explanation). Should fix the transfer failure @nahuhh reported [here](https://github.com/seraphis-migration/monero/pull/39#issuecomment-2970520570) (@nahuhh you'd need to include #39 and this PR to transfer successfully).

Credit to @nahuhh for triggering this with solid testing and maintaining a reproducible setup that made it easier to track down and fix.

### CODE REVIEW

**Comment by nahuhh** [comment]
Success 🔥

File: src/fcmp_pp/curve_trees.cpp [hunk 1]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
-    // helios scalars from selene points
-    std::vector<std::vector<fcmp_pp::HeliosScalar>> helios_scalars;
-    for (const auto &selene_points : path.c1_layers)
+    const size_t n_tree_layers = path.c1_layers.size() + path.c2_layers.size();
+    const bool root_is_c1 = n_tree_layers % 2 == 1;
+    const size_t n_c2_scalar_chunk_layers = path.c1_layers.size() - size_t(root_is_c1);
+    const size_t n_c1_scalar_chunk_layers = path.c2_layers.size() - size_t(!root_is_c1);
+
+    // c2 helios scalars from c1 selene points
+    std::vector<std::vector<fcmp_pp::HeliosScalar>> c2_scalar_chunks;
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp:1540]:
```suggestion
    std::vector<std::vector<fcmp_pp::HeliosScalar>> c2_scalar_chunks;
    c2_scalar_chunks.reserve(n_c2_scalar_chunk_layers);
```

File: src/fcmp_pp/curve_trees.cpp [hunk 2]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
-      auto &helios_layer = helios_scalars.back();
-      helios_layer.reserve(selene_points.size());
-      for (const auto &c1_point : selene_points)
-        helios_layer.emplace_back(m_c1->point_to_cycle_scalar(c1_point));
-      // Padding with 0's
-      for (std::size_t i = selene_points.size(); i < m_c2_width; ++i)
-        helios_layer.emplace_back(m_c2->zero_scalar());
+        const auto &c1_points = path.c1_layers.at(i);
+        c2_scalar_chunks.emplace_back();
+        auto &c2_scalar_layer = c2_scalar_chunks.back();
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp]:
```suggestion
        auto &c2_scalar_layer = tools::add_element(c2_scalar_chunks);
```

File: src/fcmp_pp/curve_trees.cpp [hunk 3]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
+        // Padding with 0's
+        for (std::size_t j = c1_points.size(); j < m_c2_width; ++j)
+            c2_scalar_layer.emplace_back(m_c2->zero_scalar());
     }
 
-    // selene scalars from helios points
-    std::vector<std::vector<fcmp_pp::SeleneScalar>> selene_scalars;
-    for (const auto &helios_points : path.c2_layers)
+    // c1 selene scalars from c2 helios points
+    std::vector<std::vector<fcmp_pp::SeleneScalar>> c1_scalar_chunks;
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp:1555]:
```suggestion
    std::vector<std::vector<fcmp_pp::SeleneScalar>> c1_scalar_chunks;
    c1_scalar_chunks.reserve(n_c1_scalar_chunk_layers);
```

File: src/fcmp_pp/curve_trees.cpp [hunk 4]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
-      auto &selene_layer = selene_scalars.back();
-      selene_layer.reserve(helios_points.size());
-      for (const auto &c2_point : helios_points)
-        selene_layer.emplace_back(m_c2->point_to_cycle_scalar(c2_point));
-      // Padding with 0's
-      for (std::size_t i = helios_points.size(); i < m_c1_width; ++i)
-        selene_layer.emplace_back(m_c1->zero_scalar());
+        const auto &c2_points = path.c2_layers.at(i);
+        c1_scalar_chunks.emplace_back();
+        auto &c1_scalar_layer = c1_scalar_chunks.back();
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp]:
```suggestion
        auto &c1_scalar_layer = tools::add_element(c1_scalar_chunks);
```

File: src/fcmp_pp/curve_trees.cpp [hunk 5]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
-      helios_layer.reserve(selene_points.size());
-      for (const auto &c1_point : selene_points)
-        helios_layer.emplace_back(m_c1->point_to_cycle_scalar(c1_point));
-      // Padding with 0's
-      for (std::size_t i = selene_points.size(); i < m_c2_width; ++i)
-        helios_layer.emplace_back(m_c2->zero_scalar());
+        const auto &c1_points = path.c1_layers.at(i);
+        c2_scalar_chunks.emplace_back();
+        auto &c2_scalar_layer = c2_scalar_chunks.back();
+        c2_scalar_layer.reserve(c1_points.size());
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp]:
```suggestion
        c2_scalar_layer.reserve(m_c2_width);
```

File: src/fcmp_pp/curve_trees.cpp [hunk 6]
```diff
@@ -1530,45 +1530,46 @@ CurveTrees<Selene, Helios>::PathForProof CurveTrees<Selene, Helios>::path_for_pr
...
-      selene_layer.reserve(helios_points.size());
-      for (const auto &c2_point : helios_points)
-        selene_layer.emplace_back(m_c2->point_to_cycle_scalar(c2_point));
-      // Padding with 0's
-      for (std::size_t i = helios_points.size(); i < m_c1_width; ++i)
-        selene_layer.emplace_back(m_c1->zero_scalar());
+        const auto &c2_points = path.c2_layers.at(i);
+        c1_scalar_chunks.emplace_back();
+        auto &c1_scalar_layer = c1_scalar_chunks.back();
+        c1_scalar_layer.reserve(c2_points.size());
```
Comment (jeffro256) [inline src/fcmp_pp/curve_trees.cpp]:
```suggestion
        c1_scalar_layer.reserve(m_c1_width);
```

**Review by jeffro256 — CHANGES_REQUESTED** [review]
Thank you for working on this! And thanks @nahuhh for repro'ing
